### ISSUE TITLE
Shadowing of top level extension methods

### ISSUE THREAD

**Issue by d-albrecht** [lampepfl/dotty#16284]
## Compiler version

Observed locally with version 3.1.3 and in scastie with any version up to version 3.2.1-RC2 (which - at the time of writing - is the latest usable version in scastie).

## Minimized code

```Scala
trait T1 {}
extension (t1: T1) {
  def doIt: Unit = println("Hello!")
}

object Definition {
  trait T3 {}
  extension (t3: T3) {
    def doIt: Unit = println("Hello!")
  }
}

object WorksAlways {
  import Definition._
  trait T2 { def getInner: T3 = new T3 {} }
  extension (t2: T2) {
    def doIt: Unit = t2.getInner.doIt
    def doStuff: Unit = t2.getInner.doIt
  }
}

object Works {
  trait T2 { def getInner: T1 = new T1 {} }
  extension (t2: T2) {
    def doStuff: Unit = t2.getInner.doIt
  }
}

object Fails {
  trait T2 { def getInner: T1 = new T1 {} }
  extension (t2: T2) {
    def doIt: Unit = t2.getInner.doIt
  }
}
```
[Scastie version](https://scastie.scala-lang.org/pLU5PIQbTkyGjFoBmgOoSA)

## Output

The definition of the extension method ``Fails.T2.doIt`` doesn't compile because it can't find ``T1.doIt``. This is due to ``Fails.T2.doIt`` shadowing ``T1.doIt``.

This is never an issue if:
- the base definition is wrapped by some type definition. ``WorksAlways`` relies on a copy of T1 with the only differences being that this copy is renamed (to avoid ambiguity) and wrapped in some object
- the new extension method uses a different name (see ``Works.T2.doStuff``)

But if:
- an extension method is defined in top level (or only inside a package object for that matter)
- is imported into another context (either explicitly or implicitly due to nested namespaces)
- where there is a identically named extension method in this new context (on a different type)

then the compiler can not find the top level extension method (at least in my example) and fails to compile the code.

## Expectation

Extension methods should be importable even if they are defined in top level (or at least if in top level in some package, I could see an argument why definitions in top level of the default package might not be importable). Especially if the ability for placing most definitions in top level is [one argument](https://docs.scala-lang.org/scala3/reference/dropped-features/package-objects.html) to deprecate package objects (which wouldn't solve the problem here, but that's beside the point),

Just because two extension methods use the same name, they shouldn't interfere with one another if they are defined on/for different types.

**Comment by soronpo** [comment]
Not a bug, but yeah not intuitive either. See:
https://contributors.scala-lang.org/t/change-shadowing-mechanism-of-extension-methods-for-on-par-implicit-class-behavior/5831

**Comment by d-albrecht** [comment]
TL;DR: I beg to differ that this isn't a bug, as this would imply that the observed behavior is in fact an intended feature. And I see no reason to assume that someone would actually wanted it that way.

-----
I tried to minimize my code snippet as much as possible while still preserving the problem. I could have provided the code segments in this format also:
Imagine a source file ``File.scala`` with content:
````
package packagename
package scope
trait T1 {}
extension (t1: T1) {
  def doIt: Unit = println("Hello!")
}
````
or alternatively:
````
package packagename
object Scope {
  trait T1 {}
  extension (t1: T1) {
    def doIt: Unit = println("Hello!")
  }
}
````
The former you would use with ``import packagename.scope._`` while the latter would be included with ``import packagename.Scope._`` but they would behave differently, where in the latter form shadowing wouldn't be an issue, while in the former format it would.
But what is the difference of on object to a package? I know, there are some nuances (if objects are used as singletons for example), but essentially they are scopes for other definitions. In that sense, an object is a namespace and a package is a namespace. Why does using a wildcard import for one scope differ from a wildcard import of the other scope construct? When internally both constructs will more or less be translated to static methods of some sort.
The only thing here is that this "bug" enforces some - as far as I can tell - unnecessary rules that technically shouldn't pose any benefits (you are encouraged to create additional objects just to encapsule some definitions, meaning that the import will feature a capital-letter word instead of a lower-case word, if using an object can replace the similar named (sub-)package).

With all due respect for Martin Odersky and what he did for Scala, but if regular methods can have the same problem, then to me that is just an indication that Scala can't properly handle method overloading **in general**. For a long time I was frustrated to see Scala being based on the JVM because of all the legacy aspects that are still there for backwards-compatibility resulting in phenomena like type erasure. With the tasty-format Scala apparently overcame the issue with this variance of method overloading. Only to introduce another similar problem with extension methods now?
For one, I'm totally fine if the compiler is given the option to give up figuring out what the developer wanted if several definitions collide / are ambiguous. It's not pretty to being forced to provide extra hints to the compiler. But as far as I know, in my example I couldn't even provide hints to help the compiler.
The other point is that I can live with a "tiered" approach to try to match function calls to all the function definitions that are in scope, where some definitions are tested first before even considering lower-tier definitions. But the problem is that you can't rely on short-cut evaluation here. Would you say that the expression ``true && false`` was ``true``, just because the first operand is ``true``? No, because short-cut evaluation for boolean and-expressions tells you, that you can only discard the latter parts if you encounter a ``false``. Likewise, you can't stop trying other function definitions just because you found a method with the right name. Only an applicable (and type-checked) function definition can be the condition to short-cut the method matching. In that way Scala can't do method overloading, if using the same name alone can cause the compiler to give up.

**Comment by d-albrecht** [comment]
Okay, I just revisited the scala docs about dropped (or soon-to-be dropped) features in scala 3, especially the article about package objects. In there is a note that:
> If several top-level definitions are overloaded variants with the same name, they must all come from the same source file.

Which - based on my understanding of compilers and language design - is probably the culprit here. I'm not exactly sure why the dev team decided that overloading methods in several sources shouldn't be possible. I suspect that it was tried but posed a problem too large to overcome in reasonable time. Just like the type erasure thing took a long time to solve. Otherwise this would seem like a rather arbitrary limitation if it wasn't for technical reasons.

That this "problem" now re-appears in the form of my example is rather unfortunate, I'd say. Yes, I still believe that this is a limitation to how scala handles method overloading and I would still hope that some day the situation would get improved. But I fear that a quick fix isn't possible here as the way of how scala links function calls as a whole could be the issue here. If it wasn't possible to parse overloaded functions from different source files, no-one would tackle this just to please me, and if overloaded functions weren't a problem, then my problem here likely wouldn't exist as well.
But... At least the compiler message is not appropriate and/or helpful for anyone who runs into this to detect and resolve the issue at hand. I abstracted the core problem a bit for this bug report but I ran into this exact issue in some code and if all the compiler can tell you is that type A was found but type B was expected, when I was the sole programmer of both the function definition and the call site and I knew that a variant of this method with the correct "argument type"/"base type" (it's an extension method after all, so the nomenclature is a bit tricky here) existed and theoretically should be in scope / correctly imported, then you can get lost/frustrated very quickly.

---

TL;DR:
- please fix that eventually, but it *could* be a more fundamental thing than just some form of short-cut evaluation like I said before.
- in the meantime, could you improve the situation with the error message? Though, I fear that changing the error message would be just as difficult as to fix the issue in the first place.
- but at least document this problem somewhere so that users (who read the documentation) can know about this limitation.
- sorry if my previous reply might have been a bit aggressive, it wasn't meant that way (but I was told that I have a tendency to sound rude at times)

**Comment by ornicar** [comment]
Is this the same issue [I just faced with 3.2.1](https://github.com/ornicar/scalabug/blob/main/extension-ambiguous/code.scala)?

```scala
object A:
  extension [T](x: Option[T]) def foo = 1

object B:
  extension [T](x: List[T]) def foo = 3

import A.*
import B.*

val _ = Option(1).foo // error: value foo is not a member of Option[Int].

/*
[info] compiling 1 Scala source to /home/thib/scalabug/extension-ambiguous/target/scala-3.2.1/classes ...
[error] -- [E008] Not Found Error: /home/thib/scalabug/extension-ambiguous/code.scala:10:18
[error] 10 |def f = Option(1).foo
[error]    |        ^^^^^^^^^^^^^
[error]    |      value foo is not a member of Option[Int].
[error]    |      An extension method was tried, but could not be fully constructed:
[error]    |
[error]    |          B.foo[T](Option.apply[Int](1))    failed with
[error]    |
[error]    |              Reference to foo is ambiguous,
[error]    |              it is both imported by import A._
[error]    |              and imported subsequently by import B._
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
 */
```
I was quite surprised by this behaviour, and found that it made the migration from implicit classes a bit tricky.

**Comment by d-albrecht** [comment]
Well, that's a different error message than the one I got, but it just underlines that scala is lacking a certain consistency here. If you overload a (package) top level method in different source files, the definition side apparently won't even compile. If you encapsulate a method in an object and overload this elsewhere, then it works just fine. And if you do whatever you did here (to be honest, I don't have made myself familiar with this whole "no curly brackets-style", so I'm not hundred percent sure how that would compile/compare to with bracketed code) the definition side is fine, but the method call fails to compile with the compiler reporting ambiguous methods (that almost sound like a type erasure error message).

OR: with the old implicits there was this issue that if you tried to import two values that could both be used as an implicit function parameter then - it I remember correctly - this wouldn't want to compile as well. So, perhaps your problem is related to this one?? Not sure though.

**Comment by som-snytt** [comment]
This is ambiguous:
```

def f = 42

object X:
  def f = 27

class C:
  import X.*
  def c = f

@main def test = println:
  C().c
```
the error
```
9 |  def c = f
  |          ^
  |          Reference to f is ambiguous.
  |          It is both defined in package <empty>
  |          and imported subsequently by import X._
```
but `WorksAlways` has the same lexical structure. Maybe the "slight modification of the name resolution rules" for extensions accounts for the difference?

The failure case is the same as
```
object AlsoFails:
  def doIt(i: Int): Unit = doIt(null.asInstanceOf[T1])
```

That is, extensions are ordinary methods subject to the usual shadowing rules, except when imported, for which there is special "overloading" treatment.
